osh/hub/scripts/osh-resalloc

```diff
@@ -0,0 +1,162 @@
+    try:
+        if scp_exit_status != 0:
+            logger.error(f"Failed to copy worker.conf to new worker {kobo_worker.name}")
+            return
+    finally:
+        # Delete the worker configs from hub
+        temporary_file.close()
+        # TODO: Shall we delete the worker here?
+
+    logger.debug(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo mv worker.conf /etc/osh/worker.conf && sudo chown root:root /etc/osh/worker.conf'")
+    # /etc/osh/hub/id_rsa.worker
+    move_worker_conf_status = os.system(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo mv worker.conf /etc/osh/worker.conf && sudo chown root:root /etc/osh/worker.conf'")
+    if move_worker_conf_status != 0:
+        logger.error(f"Failed to move worker.conf to new worker {kobo_worker.name}")
+        # TODO: Shall we delete the worker here?
+        return
+    # Start the worker after this is finished
+    logger.debug(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo systemctl restart osh-worker'")
+    start_worker_status = os.system(f"ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ConnectTimeout=20 -i {settings.RESALLOC_WORKER_SSH_PRIVATE_KEY} centos@{kobo_worker.name} 'sudo systemctl restart osh-worker'")
+    if start_worker_status != 0:
+        logger.error(f"Failed to start worker worker at {kobo_worker.name}")
+        # TODO: Shall we delete the worker here?
+        return
+
+
+def create_worker(ip_address):
+    # Worker name should be set to hostname of the worker
+    worker_name = get_hostname_from_ip(ip_address)
+    kobo_worker = create_worker_in_kobo(worker_name)
+    start_worker(kobo_worker)
+
+
+def delete_worker(ip_address):
+    from kobo.hub import models
+
+    # Worker name should be set to hostname of the worker
+    worker_name = get_hostname_from_ip(ip_address)
+    kobo_worker = models.Worker.objects.get(name=worker_name)
+    kobo_worker.delete()
+
+
+def workers_needed():
+    # This number should not exceed `max` field in the `/etc/resallocserver/pools.yaml`
+    # on the resalloc server.
+    MAX_SINGLE_USE_WORKERS = 8
+
+    import kobo.hub.models.Task
+
+    # Tasks in opened or assigned state
+    running_tasks = kobo.hub.models.Task.objects.running().count()
+
+    # Tasks waiting to be assigned
+    free_tasks = kobo.hub.models.Task.objects.free().count()
+
+    running_and_free_tasks = running_tasks + free_tasks
+
+    print(min(running_and_free_tasks, MAX_SINGLE_USE_WORKERS))
```

Comment:
Note that you need to wait some nontrivial time till the agent becomes available. So if we combine single-task workers and multi-task workers, some of them might get allocated and never used eventually.

My original thinking was that we could have something like a logarithmic function of free tasks in the queue (or something like that).  I.e. something like `ceil(ln(free_tasks)) + running_tasks`.

Reply:
Perhaps just `ceil(ln(free_tasks))` and do not return `running_tasks`.  If there are no more free tasks, there's no need to preallocate more agents when the old agents are terminated.

Reply:
I understand that allocation of machines takes time but how is it related to the number of **needed** workers?

Reply:
Extreme example:
```
initial state: two static reusable workers, 0 dynamic (agent spawner), two tasks to be processed
1. the two tasks are taken by static workers, we report "ideal number = 2"
2. agent spawner allocates the two workers
3. tasks are processed, "ideal number = 2"
4. agent spawner downsizes... and terminates agents
```

But even without static workers, if we make the dynamic workers reusable:
```
initial state: 1 dynamic worker allocated, one running task, two tasks in queue
1. ideal number reported = 3
2. spawner starts allocating 2 more agents
3. the running task is finished, the worker is free -> and takes another task
4. the two agents got allocated && ready, there's only one task to finish
```

> I understand that allocation of machines takes time but how is it related to the number of needed workers?

Since the spawner only poorly delivers the requested number of agents that OSH asks for, we can estimate or guess in advance whether the workers will be needed. No big deal, the stack of workers will get downsized anyway... but there's unnecessary allocation/deallocation traffic.

If the particular OSH instance uses only dynamic && single-task workers, we don't have to worry. Combining with static workers is probably not considered in the public OSH instance anyway. However, reusing allocated agents seems like a perfectly valid use case; that's what we actually do in Fedora Copr, too. Builders are recycled for multiple tasks as long as it is safe to do so (the task is owned by the same user, building into the same project, etc.).

Reply:
Yes, I agree and I think you are right (originally we did not plan the spawner would exist, I think?).

Now, I think we should document this constraint, somewhere (either 1. or 2. and the details about 2.).  But note that it is rather an artificial constraint (not sure there are technical reasons to implement it - to have two types of workers, and whether OSH has to know about this at all).